be/src/util/json_flater.cpp

```diff
@@ -0,0 +1,384 @@
+    }
+    return types;
+}
+
+struct FlatColumnDesc {
+public:
+    uint8_t type = JsonFlater::JSON_NULL_TYPE_BITS;
+    uint64_t hits = 0;
+    uint16_t casts = 0;
+};
+
+void JsonFlater::derived_paths(std::vector<ColumnPtr>& json_datas) {
+    _flat_paths.clear();
+    _flat_types.clear();
+
+    if (json_datas.empty()) {
+        return;
+    }
+
+    size_t total_rows = 0;
+    size_t null_count = 0;
+
+    for (auto& column : json_datas) {
+        total_rows += column->size();
+        if (column->only_null() || column->empty()) {
+            null_count += column->size();
+            continue;
+        } else if (column->is_nullable()) {
+            auto* nullable_column = down_cast<NullableColumn*>(column.get());
+            null_count += nullable_column->null_count();
+        }
+    }
+
+    // more than half of null
+    if (null_count > total_rows * config::json_flat_null_factor) {
+        VLOG(8) << "flat json, null_count[" << null_count << "], row[" << total_rows
+                << "], null_factor: " << config::json_flat_null_factor;
+        return;
+    }
+
+    // extract common keys, type
+    std::unordered_map<std::string_view, FlatColumnDesc> derived_maps;
+    for (size_t k = 0; k < json_datas.size(); k++) {
+        size_t row_count = json_datas[k]->size();
+
+        ColumnViewer<TYPE_JSON> viewer(json_datas[k]);
+        for (size_t i = 0; i < row_count; ++i) {
+            if (viewer.is_null(i)) {
+                continue;
+            }
+
+            JsonValue* json = viewer.value(i);
+            auto vslice = json->to_vslice();
+
+            if (vslice.isNull() || vslice.isNone() || vslice.isEmptyObject() || !vslice.isObject()) {
+                continue;
+            }
+
+            vpack::ObjectIterator iter(vslice);
+            for (const auto& it : iter) {
+                std::string_view name = it.key.stringView();
+                derived_maps[name].hits++;
+                uint8_t type = derived_maps[name].type;
+                uint8_t compatibility_type =
+                        JsonFlater::get_compatibility_type(it.value.type(), derived_maps[name].type);
+                derived_maps[name].type = compatibility_type;
+                derived_maps[name].casts += (type != compatibility_type);
+            }
+        }
+    }
+
+    if (derived_maps.size() <= config::json_flat_internal_column_min_limit) {
+        VLOG(8) << "flat json, internal column too less: " << derived_maps.size()
+                << ", at least: " << config::json_flat_internal_column_min_limit;
+        return;
+    }
+
+    // sort by hit, casts
+    std::vector<pair<std::string_view, FlatColumnDesc>> top_hits(derived_maps.begin(), derived_maps.end());
+    std::sort(top_hits.begin(), top_hits.end(),
+              [](const pair<std::string_view, FlatColumnDesc>& a, const pair<std::string_view, FlatColumnDesc>& b) {
+                  // check hits, the higher the hit rate, the higher the priority.
```

Comment:
what's the hit meaning?

Reply:
we can extract columns with a certain proportion(<= 30%) of null/none values, but columns with a lower proportion of null/none values have higher priority.

Reply:
add comment